### DISCUSSION TITLE
AWS배포시 요청 URL undefined와 @material-ui 라이브러리 문제 문의드립니다

### DISCUSSION THREAD

**Discussion by lutae2000** [fsoftwareengineer/todo-application#71] [질문과 답변] [answered]
안녕하세요 작가님!

책 내용을 따라가던중 아래와같이 두가지 이슈가 있습니다

1. AWS에 Elastic beanTalk 배포후  로그인, 회원원가입, 글쓰기등 요청 URL 중간에 undefined가 껴있어서 오류가 발생합니다.
올려주신 소스와 책 내용을 더블 체크 했으나 특이점을 발견할수 없습니다
=> 배포된 동일한 소스로 로컬 개발환경 <> AWS RDS로 구동시엔 전혀 문제가 없음
![image](https://user-images.githubusercontent.com/40047335/150681598-3b7f28fd-7ee0-45a6-884a-466cd36ad9a6.png)

api-service.js 파일 내 요청 URL
![image](https://user-images.githubusercontent.com/40047335/150683200-08151c63-b2a9-4eac-8b81-a4f7213f30cd.png)

2. npm run build로  빌드시 @material-ui/core 가 빠져서 EC2서버에서 수동으로 받아야하는데 이부분은 원래 그런건가요...?
=> @marerial-ui/icon는 존재

**Comment by fsoftwareengineer** [comment, marked as answer]
ApiService.js 와 app-config.js 코드를 댓글에 남겨주시면 확인 해 보도록 하겠습니다. 

material-ui/core의 경우 package.json이 아래와 같이 dependencies아래에 존재하는지 확인 해 주시기 바랍니다. 

```
{
  "name": "todo-react-app",
  "version": "0.1.0",
  "private": true,
  "dependencies": {
    "@material-ui/core": "^4.11.3",
    "@material-ui/icons": "^4.11.2",
    "@testing-library/jest-dom": "^5.11.9",
    "@testing-library/react": "^11.2.5",
    "@testing-library/user-event": "^12.8.1",
    "react": "^17.0.1",
    "react-dom": "^17.0.1",
    "react-router-dom": "^5.2.0",
    "react-scripts": "4.0.3",
    "web-vitals": "^1.1.0"
  },
```

**Reply by lutae2000** [reply]
안녕하세요 작가님! 
답변 감사합니다
package.json 내 dependencies가 빠져있었네요;; 이것 때문에 한참을 헤맸는데 감사합니다!

ApiService.js 와 app-config.js 코드 첨부 드립니다!


apiService.js

```
import { API_BASE_URL } from "./api-config";
const ACCESS_TOKEN = "ACCESS_TOKEN";

export function call(api, method, request) {
  const accessToken = localStorage.getItem("ACCESS_TOKEN");

  let headers = new Headers({
    "Content-Type": "application/json",
  });

  if (accessToken && accessToken != null) {
    headers.append("Authorization", "Bearer " + accessToken);
  }

  let options = {
    headers: headers,
    url: API_BASE_URL + api,
    method: method,
  };

  if (request) {
    //Get method
    options.body = JSON.stringify(request);
  }

  return fetch(options.url, options)
    .then((response) =>
      response.json().then((json) => {
        if (!response.ok) {
          // response.ok가 true이면 정상적인 리스폰스를 받은것, 아니면 에러 리스폰스를 받은것.
          return Promise.reject(json);
        }
        return json;
      })
    )
    .catch((error) => {
      console.log(error.status);
      if (error.status === 403) {
        window.location.href = "login"; //코드403 받으면 login 페이지로 던짐
      }
      return Promise.reject(error);
    });
}

export function signin(userDTO) {
  return call("/auth/signin", "POST", userDTO).then((response) => {
    console.log("response: " + response);

    if (response.token) {
      localStorage.setItem("ACCESS_TOKEN", response.token);
      window.location.href = "/";
    }
  });
}

export function signout() {
  localStorage.setItem(ACCESS_TOKEN, null);
  window.location.href = "/login";
}

export function signup(userDTO) {
  return call("/signup", "POST", userDTO);
}
```


api-config.js

```
let backendHost;

const hostname = window && window.location && window.location.hostname;

if (hostname === "localhost") {
  backendHost = "http://localhost:8080";
} else {
 backendHost = [백엔드 호스트 DNS]
}

export const API_BASE_URL = `${backendHost}`;
```

**Reply by fsoftwareengineer** [reply]
제 생각에는 API_BASE_URL이 제대로 설정되지 않은 것 같습니다. API_BASE_URL 이 undefined여서 default url인 현재 url(프론트엔드 url)에 undefined/auth/signin이렇게 들어가는것 같은데, backendHost가 잘 설정되어 있는지 signin 함수 내부에서 console.log를 찍어 확인 해 보면 좋을 것 같습니다.

**Reply by fsoftwareengineer** [reply]
api-config와 ApiService.js는 같은 디렉터리에 존재하나요?

**Reply by lutae2000** [reply]
> api-config와 ApiService.js는 같은 디렉터리에 존재하나요?

네 같은 디렉토리에 존재합니다

답변해주신데로 API_BASE_URL 콘솔 로그 찍어서 확인 해보겠습니다
바쁘신와중에 답변 감사드립니다!

**Reply by lutae2000** [reply]
문제점을 찾았습니다만 프론트엔드를 수정후 재 배포해도 수정된 내역이 엘라스틱 빈톡에 정상 반영이 안되는것 같습니다
백엔드 프로젝트처럼 수정한 내역 빌드 후
```
$ eb deploy
```
를 입력해도 반영이 안되는데 프론트엔드쪽 수정한 내역을 엘라스틱빈톡에 반영하는 방법이 다른지 궁금합니다

개발자도구 => 소스에서 확인
![image](https://user-images.githubusercontent.com/40047335/150798228-775ad0d9-5c53-4da3-afdd-7673e687b63b.png)

**Reply by fsoftwareengineer** [reply]
혹시 소스코드 관리를 git으로 하신다면 commit후 eb deploy를 다시 해야합니다.

**Reply by lutae2000** [reply]
늦은시간에 답변 감사합니다

가이드 해주신데로 git commit & push 후 eb deploy하니 정상적으로 반영되었습니다!
수정내역 배포가 안되어서 계속 어떤게 문제인지 찾고 있었는데 git이 영향이 있을줄은 몰랐네요;;
